Convert singular polynomials to AbstractAlgebra polynomials

heiderich opened the pull request:
These functions allow converting singular polynomial rings and polynomials to polynomial rings and polynomials in AbstractAlgebra.

The name of the existing function

    PolynomialRing(R::AbstractAlgebra.Generic.MPolyRing{T}; cached::Bool = true, ordering::Symbol = :degrevlex, ordering2::Symbol = :comp1min, degree_bound::Int = 0)  where {T <: RingElement}

was changed to

    AsEquivalentSingularPolynomialRing

The "converse" is the function

    function AsEquivalentAbstractAlgebraPolynomialRing(R::Singular.PolyRing{Singular.n_unknown{T}}; ordering::Symbol = :degrevlex)  where {T <: Nemo.RingElem}

Note however that after

```
using Cxx;
using Singular;
K = Nemo.ArbField(64);
R,vars_R = AbstractAlgebra.Generic.PolynomialRing(K, ["x","y"]);
RR,vars_RR = Singular.AsEquivalentSingularPolynomialRing(R);
RRR,vars_RRR = Singular.AsEquivalentAbstractAlgebraPolynomialRing(RR)
```

the output of 

`RRR == R`

is `false`, as a new ring is constructed. In my opinion this is the desired behavior.

wbhart commented:
I don't think the convert function in n_unknown.jl will work. You are doing an unsafe load, and if the gc cleans anything up in the mean time, the returned pointer could be invalid. We have functions "number" and "julia" which usually do "conversions" of this kind, though I don't exactly know what it is being used for here.

In any event, I don't think we should be using convert anywhere, since it implies that things can be converted only knowing their types, which is very rare.

wbhart commented:
At the moment the diff shows that one of the PolynomialRing functions is removed by the PR. Is that intentional?

heiderich commented:
Yes, I removed 

```
function PolynomialRing(R::AbstractAlgebra.Generic.MPolyRing{T}; cached::Bool = true,
      ordering::Symbol = :degrevlex, ordering2::Symbol = :comp1min,
      degree_bound::Int = 0)  where {T <: RingElement}
```

and replaced it by

```
function AsEquivalentSingularPolynomialRing(R::AbstractAlgebra.Generic.MPolyRing{T}; cached::Bool = true,
      ordering::Symbol = :degrevlex, ordering2::Symbol = :comp1min,
      degree_bound::Int = 0)  where {T <: RingElement}
```

wbhart commented:
I see. Can you add a docstring for it and reference it from the docs.

wbhart commented:
Perfect, thanks. Once the tests pass I'll merge this.

heiderich commented:
They failed. But most of the recent commits on the master branch too.

wbhart commented:
Only the apple one needs to pass at present.
